File: native/swift/Sources/Deview/Palette.swift
import CDeview
import CoreGraphics

/// A plain C enum arrives in Swift as a struct whose rawValue is unsigned, while every field that
/// holds one across the ABI is `int32_t`. These do the conversion once rather than at every
/// comparison.
extension DeviewRowKind {
    var value: Int32 { Int32(rawValue) }
}

extension DeviewKey {
    var value: Int32 { Int32(rawValue) }
}

extension DeviewQueueFlags {
    var value: Int32 { Int32(rawValue) }
}

extension DeviewButtonFlags {
    var value: Int32 { Int32(rawValue) }
}

/// Transcribed from `RowColour` and `RowBackground` in deview.cpp, so a change looks the same
/// whichever renderer drew it. The screen model carries a row kind and never a colour, so this
/// mapping belongs to each renderer rather than to the model.
enum Palette {
    static let background = grey(24)
    static let filler = grey(28)
    static let text = grey(212)

    /// The gutter, and the subtitle and status that the other heads draw dimmed.
    static let dim = grey(130)

    static let rule = grey(70)

    /// ImGui draws a selected item as its accent at 31% over the window background. This is that
    /// composite, so the queue highlight matches without carrying an alpha channel around.
    static let selected = rgb(38, 64, 90)

    static let buttonFace = grey(52)
    static let buttonDisabled = grey(34)

    static func foreground(_ kind: Int32) -> CGColor {
        switch kind {
        case DEVIEW_ROW_ADDED.value:
            return rgb(126, 214, 139)
        case DEVIEW_ROW_REMOVED.value:
            return rgb(233, 129, 129)
        case DEVIEW_ROW_MODIFIED.value:
            return rgb(231, 197, 113)
        default:
            return text
        }
    }

    /// Nil where the row takes the window background, which is every unchanged row.
    static func rowBackground(_ kind: Int32) -> CGColor? {
        switch kind {
        case DEVIEW_ROW_ADDED.value:
            return rgb(38, 74, 44)
        case DEVIEW_ROW_REMOVED.value:
            return rgb(84, 40, 40)
        case DEVIEW_ROW_MODIFIED.value:
            return rgb(74, 64, 32)
        case DEVIEW_ROW_FILLER.value:
            return filler
        default:
            return nil
        }
    }

    static func marker(_ kind: Int32) -> String {
        switch kind {
        case DEVIEW_ROW_ADDED.value:
            return "+"
        case DEVIEW_ROW_REMOVED.value:
            return "-"
        case DEVIEW_ROW_MODIFIED.value:
            return "~"
        default:
            return " "
        }
    }

    private static func rgb(_ red: Int, _ green: Int, _ blue: Int) -> CGColor {
        CGColor(
            srgbRed: CGFloat(red) / 255,
            green: CGFloat(green) / 255,
            blue: CGFloat(blue) / 255,
            alpha: 1)
    }

    private static func grey(_ level: Int) -> CGColor {
        rgb(level, level, level)
    }
}

File: src/DiffEngineViewer.Windows/Palette.cs
/// <summary>
/// Transcribed from <c>RowColour</c> and <c>RowBackground</c> in the native shim, so a change looks
/// the same whichever renderer drew it. The <see cref="Screen" /> model deliberately carries a
/// <see cref="RowKind" /> and never a colour, so this mapping is each renderer's own.
/// </summary>
static class Palette
{
    public static readonly Color Background = Color.FromArgb(24, 24, 24);
    public static readonly Color Filler = Color.FromArgb(28, 28, 28);
    public static readonly Color Text = Color.FromArgb(212, 212, 212);

    /// <summary>
    /// The gutter, and the subtitle and status that ImGui draws with TextDisabled.
    /// </summary>
    public static readonly Color Dim = Color.FromArgb(130, 130, 130);

    public static readonly Color Rule = Color.FromArgb(70, 70, 70);

    /// <summary>
    /// The two squares behind a picture. Dark enough not to compete with the image, and different
    /// enough from each other that a transparent region is obviously transparent rather than
    /// just flat.
    /// </summary>
    public static readonly Color CheckerLight = Color.FromArgb(64, 64, 64);

    public static readonly Color CheckerDark = Color.FromArgb(48, 48, 48);

    /// <summary>
    /// ImGui draws a selected Selectable as its accent at 31% over the window background. This is
    /// that composite, so the queue highlight matches without carrying an alpha channel around.
    /// </summary>
    public static readonly Color Selected = Color.FromArgb(38, 64, 90);

    /// <summary>
    /// Behind selected pane text. Brighter than the queue's row highlight so the two do not read
    /// as the same thing, and dark enough that added green and removed red are still legible over
    /// it - the text keeps its own colour when selected, because what kind of change a line is
    /// does not stop mattering while it is being copied.
    /// </summary>
    public static readonly Color Selection = Color.FromArgb(55, 92, 130);

    public static Color Foreground(RowKind kind) =>
        kind switch
        {
            RowKind.Added => Color.FromArgb(126, 214, 139),
            RowKind.Removed => Color.FromArgb(233, 129, 129),
            RowKind.Modified => Color.FromArgb(231, 197, 113),
            _ => Text
        };

    /// <summary>
    /// Null where the row takes the window background, which is every unchanged row.
    /// </summary>
    public static Color? RowBackground(RowKind kind) =>
        kind switch
        {
            RowKind.Added => Color.FromArgb(38, 74, 44),
            RowKind.Removed => Color.FromArgb(84, 40, 40),
            RowKind.Modified => Color.FromArgb(74, 64, 32),
            RowKind.Filler => Filler,
            _ => null
        };

    public static char Marker(RowKind kind) =>
        kind switch
        {
            RowKind.Added => '+',
            RowKind.Removed => '-',
            RowKind.Modified => '~',
            _ => ' '
        };
}
